bin/nodetool-wrapper

```
@@ -0,0 +1,29 @@
+#!/bin/bash
+#
+# Copyright (C) 2023-present ScyllaDB
+#
+
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+SCRIPT_PATH="$(dirname $(realpath $0))"
+
+# nodetool is installed to $prefix/share/cassandra/bin, while scylla to $prefix/bin
+SCYLLA_PATH="$(realpath $SCRIPT_PATH/../../../bin/scylla)"
+NODETOOL_PATH="$(realpath $SCRIPT_PATH/nodetool.java)"
+
+# Capture the output of scylla-nodetool, as it might just print an error about
+# the command not existing.
```

An idea how to improve this (I think I mentioned this in the past, it's not important to do this now but I think it will be a good idea for you to add alter) -

I think you should add a command line option, e.g., "-q" (quiet) or "-s" (silent), which would cause "scylla nodetool" to be silent if the command is not supported and JUST return the error code 100. This will allow you to get rid of this ugly redirection trick, which isn't just ugly (sorry), it's also partially wrong because it:
1. Can buffer output longer than scylla-nodetool wanted to.
2. Will combine both stderr to stdout into one stdout stream.

---

Right - I knew I thought about this already :-)
I prefer my proposed solution ("-q") over your "--help" one, because in the optimal case (90% of the nodetool runs will just use Scylla) it will just run scylla once, not twice.

---

If this was a longer-term solution, I would prefer the `-q/-s` option. But since I plan to wrap the transition up in a few months, I don't want to add an option, which doesn't make sense outside the scope of this script (a user always wants readable error messages, instead of silent failures).
We can live with the extra `--help` call for this transitional period.

---

Yes, we will have this script for some time, unless we backport scylla-nodetool patches (which I doubt).
Note that implementing this `-q` would not be trivial. We lookup the command and check for `--help` very early in the startup phase. At this point, the regular arguments are not even parsed yet. So I'd have to shuffle the startup code around to add support to this special `-q` flag and the only thing it would change is that I wouldn't need a redirection to `/dev/null`, which is not a large gain.

As for the implementation state, currently, we stand at 31/48 commands, so 17 more commands to go. I already implemented all the low-hanging fruits, still I think I will likely finish by the 6.0 timeline (although I don't promise).

---

I implemented @tchaikov's suggestion in v2. It indeed makes the code much nicer.

---

I just realized that with `-q` we would only need to run `scylla-nodetool` once, instead of twice. I don't think this makes a big difference, `scylla-nodetool` starts very fast, and the slow path (scylla-nodetool + java-nodetool) still remains, with both solutions.

---

I'm not sure what you mean. With "-q" you run it just once - it either does the right thing, or in the case of missing support (and only in that case) it fails **silently**. I think with @tchaikov's "--help" suggestion you need to run it twice - once just to check the validity, and again to run normally. This is why I liked this idea less. But I agree that if scylla-nodetool starts very fast (surprising, considering it is probably a huge executable which uses Seastar) it doesn't matter.